Sources/JSExtension.swift

```diff
@@ -0,0 +1,338 @@
+    /// and can return any object. 
+    ///
+    /// - Parameters:
+    ///   - name: filter's name
+    ///   - code: JavaScript code for filter
+    ///   - jsContext: JavaScript context to use for code execution. Uses extension's context by default
+    ///
+    /// Example:
+    /// ```
+    /// function uppercase(value, params) {
+    ///   return value.toUpperCase()
+    /// }
+    /// ```
+    public func registerFilter(_ name: String, script code: String) {
+        self.registerFilter(name, filter: { [unowned self] (value: Any?, params: [Any?]) throws -> Any? in
+            guard let value = value else { return nil }
+            try inJSContext(self.jsContext) { self.jsContext.evaluateScript(code) }
+            let filter = try inJSContext(self.jsContext) { self.jsContext.objectForKeyedSubscript(name) }
+            let result = try inJSContext(self.jsContext) { filter?.call(withArguments: [value, params]) }
+            return result?.toObject()
+        })
+    }
+
+    /// Registers JavaScript function as a simple tag. Function should have the same name as `name` parameter,
+    /// should accept singe parameters for context and should return a string.
+    ///
+    /// - Parameters:
+    ///   - name: filter's name
+    ///   - code: JavaScript code for tag
+    ///   - jsContext: JavaScript context to use for code execution. Uses extension's context by default.
+    ///
+    /// Example:
+    /// ```
+    /// function greet(context) {
+    ///   return \"Hello, \" + context.valueForKey('name')
+    /// }
+    /// ```
+    public func registerSimpleTag(_ name: String, script code: String) {
+        self.registerSimpleTag(name) { [unowned self] (context) -> String in
+            try inJSContext(self.jsContext) { self.jsContext.evaluateScript(code) }
+            let tag = try inJSContext(self.jsContext) { self.jsContext.objectForKeyedSubscript(name) }
+            let result = try inJSContext(self.jsContext) { tag?.call(withArguments: [JSStencilContext(context)]) }
+            return result?.toString() ?? ""
+        }
+    }
+
+    /// Registers JavaScript function as a tag. Function should have the same name as `name` parameter,
+    /// should be constructed with two parameters for parser and token and should define function `render`
+    /// to render its content that should accept single parameter for context and should return string.
+    ///
+    /// - Parameters:
+    ///   - name: filter's name
+    ///   - code: JavaScript code for tag
+    ///   - jsContext: JavaScript context to use for code execution. Uses extension's context by default.
+    ///
+    /// Example:
+    /// ```
+    /// function greet(parser, token) {
+    ///  // use parser and token to parse tag
+    ///
+    ///  this.render = function(context) {
+    ///    // return rendered content as a string
+    ///  }
+    /// }
+    /// ```
+    public func registerTag(_ name: String, script code: String) {
```

Comment:
Are all the wrappers below (such as JSTokenParser, JSToken, ...) needed for only this call? If so, we might want to re-think it. I'm not a fan of repeating and creating copies of everything in Stencil, just to support this. What if stencil adds new types, or changes how they work?

Maybe a more limited version would be preferable. For example, define a way that a javascript function/object can define a list of parameters and a render function. And then unify this with the "simple tag" you have above.